### ISSUE TITLE
FsaVec design

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#231]
I'm thinking it may be easiest to have a single class that supports both (C++) Fsa and FsaVec, which of course are actually the same type; we can call it k2.Fsa.  The .shape attribute will tell us how many axes there are, e.g. f.shape = (256, None, None) for a vector of 56 Fsa's or f.shape = (13, None) for an Fsa.  Note, this .shape doesn't exist at the C++ level but can be obtained from the .NumAxes() and .Dim0().  (Note: below, assume a variable f is of type k2.Fsa).

f.arcs will be a RaggedArc, i.e. a pybind11-wrapped version of Ragged<Arc>, and its members will be accessible as usual, e.g. f.arcs.row_ids(1), f.arcs.values and so on, although, we can declare that users should not use that interface to mutate anything.

Generally speaking the only way to *construct* a k2.Fsa will be through one of its constructors or by an FSA operation such as composition, top-sorting, etc.   (Those operations will likely be available as k2.fsa.Intersect(), and so on, which will accept k2.Fsa objects and invoke the `_k2.xxxx` functions exposed by pybind11.   Once constructed, the object's structural elements likely won't be mutable, except we may support, say, swapping the input and output labels or setting the weights (for which, see below) as long as the structure is unchanged.  Such operations would probably be enabled through special interfaces, not by having the users simply write to fields.

The Fsa will have arbitrary attributes which are PyTorch tensors and whose first dimension equals the total number of arcs (i.e. f.arcs.numel()).  These will be stored in a dict and accessed via `__getattr__` and `__setattr__`.  
   - The `__setattr__` will just check that the first index equals the arcs.numel(), and set it in the dict.  However, if it is the .weight attribute it will also overwrite the weights in the FSA with it, i.e. the .scores field, in addition to setting the class member.
  - The `__getattr__` will just return the attribute.  It will also support getting certain "special" attributes, particularly the .labels (or .symbols?).

Fsa operations will propagate the attributes as follows.  
   - Unary operations where each output-arc corresponds to one input-arc, such as top-sort:
       Just do output_attr = input_attr[arc_map]

   - Unary operations where each output-arc corresponds to zero or more input-arcs, such as
     determinization:
       - For floating-point attributes, let the output_attr be the sum of the attributes at corresponding input arcs (can be done with index_add(), I believe).
       - For integer attributes, we would let the output_attr be a ragged tensor with the input attributes appended; this wouldn't be done with PyTorch itself but we'd have to wrap some C++ code.  We can handle this later; anyway it doesn't require differentiation.

 - Binary operations where each output-arc corresponds to either inputa_arc or inputb_arc or the pair (inputa_arc,inputb_arc); I'm thinking about composition; the 1st 2 cases relate to epsilons.
      - For floating-point attributes, let the output_attr be the sum of the corresponding input attributes.
      - For integer attributes, this won't be supported.  (I.e. we won't support composition where both the input
         and output FSAs have integer attributes of the same name).
 
 - Unary operations where the output is a scalar, such as computing the total score:
     - For now we may just not do anything with these extra fields in this case, but we could perhaps optionally support adding together the floating-point attributes and appending the integer ones to a ragged tensor.
     - Note: for backprop purposes we may compute this total-score in a special way using the arc_map, using PyTorch operations, rather than fust taking the floating-point output from k2.  Depends on the semiring though.  Dont worry about this for now, I'll figure it out.

  - Note: we will always create an attribute called '.scores' when we construct an FSA, e.g. from a tensor.  This .scores attribute will be parallel to the .score field of the arcs, the idea is that they will always be the same (but separately stored).  If the user overwrites the .scores field, we will also automatically overwrite the .score elements of the arcs.  The reason for doing this is for backprop purposes: we won't have to worry about how PyTorch treats backprop when we're reinterpreting int32's as floats.  We'll be using PyTorch indexing to propagate the .scores fields when we do operations, so backprop will automagically happen without our having to do anything special.

I changed my mind about putting extra fields in a separate sub-object called per_arc.  We can just set them in the k2.Fsa object directly, it's easier to code.

**Comment by csukuangfj** [comment]
Is anyone currently fixing this issue? If not, I'll start.

**Comment by qindazhu** [comment]
I think no one currently is working on this. You can start Fangjun, many thanks!

**Comment by csukuangfj** [comment]
okay.
